x-pack/plugins/enterprise_search/server/plugin.ts

```diff
@@ -59,18 +58,19 @@ interface PluginsStart {
 
 export interface RouteDependencies {
   router: IRouter;
-  config: ConfigType;
+  config: EnterpriseSearchConfig;
   log: Logger;
   enterpriseSearchRequestHandler: IEnterpriseSearchRequestHandler;
   getSavedObjectsService?(): SavedObjectsServiceStart;
 }
 
 export class EnterpriseSearchPlugin implements Plugin {
-  private readonly config: ConfigType;
+  private readonly config: EnterpriseSearchConfig;
   private readonly logger: Logger;
 
   constructor(initializerContext: PluginInitializerContext) {
-    this.config = initializerContext.config.get<ConfigType>();
+    const rawConfig = initializerContext.config.get<ConfigType>();
+    this.config = new EnterpriseSearchConfig(rawConfig);
```

Comment:
So after going through this PR a few times, if I'm understanding correctly this is what's happening:

1. New config yml is coming in that looks like this:
```yml
  ssl:
    certificateAuthorities: ['something']
    rejectUnauthorized: true
```

2. We added a helper in the middle takes this new config data, loads the certificate authorities, and creates a non-default http agent

3. Our `EnterpriseSearchRequestHandler` helper then inserts this new http agent as an `agent:` key in our API calls to Enterprise Search

If all that is correct, and please stop me if I'm wrong, then I'd really like to push back against creating a lib that's called/focused on `enterprise_search_config.ts`. If the main outcome of this helper is a http agent, I would strongly prefer something more modular & focused like `lib/certificate_http_agent.ts` or similar.

My main issue here is the override of the `config` dependency we're passing around. The problem is that the incoming config (from Kibana) and outgoing config (from this helper) has changed, and IMO that's confusing not a good thing. When I pass in a `config` dependency to a route, I expect the config I get to be what's defined in and set Kibana - I don't expect other devs to have changed it along the way (in this case: nested ssl keys moved to top-level key, new `httpAgent` key). There are some scenarios where mutating the Kibana config _might_ be warranted, but IMO this is not one of them.

Instead of overriding our `config`, I strongly think we should prefer an architecture that's confined to `EnterpriseSearchRequestHandler` and looks something like this:

```ts
// enterprise_search_request_handler.ts
import { getHttpAgent } from './certificate_http_agent';

class EnterpriseSearchRequestHandler {
  // ...

        // Set up API options
        const options = {
          method: request.route.method as string,
          headers: { Authorization: request.headers.authorization as string, ...JSON_HEADER },
          body: this.getBodyAsString(request.body as object | Buffer),
          agent: getHttpAgent(config),
        };

        // Call the Enterprise Search API
        const apiResponse = await fetch(url, options);

  // ...
}
```

```ts
// certificate_http_agent.ts
import { readFileSync } from 'fs';

import http from 'http';
import https from 'https';

export const getHttpAgent = (config: ConfigType): http.Agent | https.Agent => {
  const defaultAgent = new http.Agent();

  if (!config.ssl) return defaultAgent;

  const { certificateAuthorities, rejectUnauthorized } = config.ssl;

  try {
    const parsedHost = new URL(config.host);
    if (parsedHost.protocol === 'https:') {
      return new https.Agent({
        ca: loadCertificateAuthorities(certificateAuthorities),
        rejectUnauthorized: !!rejectUnauthorized,
      });
    }
  } catch {
    // ignore the error
  }

  return defaultAgent;
}

// Loads custom CA certificate files and returns all certificates as an array
export const loadCertificateAuthorities(ca: string | string[]): string[] = () => {
  const parsedCerts = [];
  if (ca) {
    const paths = Array.isArray(ca) ? ca : [ca];
    if (paths.length > 0) {
      for (const path of paths) {
        const parsedCert = readFileSync(path, 'utf8');
        parsedCerts.push(parsedCert);
      }
    }
  }
  return parsedCerts;
}
```

With this approach, we avoid touching/modifying `config` whatsoever, and keep the lib limited to what it touches (our API requests to Enterprise Search). Thoughts?

Reply:
> My main issue here is the override of the config dependency we're passing around

I have followed the example of `ElasticsearchConfig`, `HttpConfig`, etc and encapsulated all of that logic in the config object since it is directly related to the settings in the config file.

Reply:
I'll make the change to move the method into a separate function loaded in appropriate places.